src/dawn/CodeGen/Cuda/CudaCodeGen.cpp

```diff
@@ -0,0 +1,1139 @@
+  }
+
+  return strides;
+}
+
+iir::Extents CudaCodeGen::computeTempMaxWriteExtent(iir::Stencil const& stencil) const {
+  auto tempFields = makeRange(
+      stencil.getFields(),
+      std::function<bool(std::pair<int, iir::Stencil::FieldInfo> const&)>(
+          [](std::pair<int, iir::Stencil::FieldInfo> const& p) { return p.second.IsTemporary; }));
+  iir::Extents maxExtents{0, 0, 0, 0, 0, 0};
+  for(auto field : tempFields) {
+    DAWN_ASSERT((*field).second.field.getWriteExtentsRB().is_initialized());
+    maxExtents.merge(*((*field).second.field.getWriteExtentsRB()));
+  }
+  return maxExtents;
+}
+void CudaCodeGen::addTempStorageTypedef(Structure& stencilClass,
+                                        iir::Stencil const& stencil) const {
+
+  auto maxExtents = computeTempMaxWriteExtent(stencil);
+  stencilClass.addTypeDef("tmp_halo_t")
+      .addType("gridtools::halo< " + std::to_string(-maxExtents[0].Minus) + "," +
+               std::to_string(-maxExtents[1].Minus) + ", 0, 0, " +
+               std::to_string(getVerticalTmpHaloSize(stencil)) + ">");
+
+  stencilClass.addTypeDef(tmpMetadataTypename_)
+      .addType("storage_traits_t::storage_info_t< 0, 5, tmp_halo_t >");
+
+  stencilClass.addTypeDef(tmpStorageTypename_)
+      .addType("storage_traits_t::data_store_t< float_type, " + tmpMetadataTypename_ + ">");
+}
+
+void CudaCodeGen::addTmpStorageInit(
+    MemberFunction& ctr, iir::Stencil const& stencil,
+    IndexRange<const std::unordered_map<int, iir::Stencil::FieldInfo>>& tempFields) const {
+  auto maxExtents = computeTempMaxWriteExtent(stencil);
+
+  const auto blockSize = stencil.getParent()->getBlockSize();
+
+  if(!(tempFields.empty())) {
+    ctr.addInit(tmpMetadataName_ + "(" + std::to_string(blockSize[0]) + "+" +
+                std::to_string(-maxExtents[0].Minus + maxExtents[0].Plus) + ", " +
+                std::to_string(blockSize[1]) + "+" +
+                std::to_string(-maxExtents[1].Minus + maxExtents[1].Plus) + ", (dom_.isize()+ " +
+                std::to_string(blockSize[0]) + " - 1) / " + std::to_string(blockSize[0]) +
+                ", (dom_.jsize()+ " + std::to_string(blockSize[1]) + " - 1) / " +
+                std::to_string(blockSize[1]) + ", dom_.ksize() + 2 * " +
+                std::to_string(getVerticalTmpHaloSize(stencil)) + ")");
+    for(auto fieldIt : tempFields) {
+      ctr.addInit("m_" + (*fieldIt).second.Name + "(" + tmpMetadataName_ + ")");
+    }
+  }
+}
+
+std::unique_ptr<TranslationUnit> CudaCodeGen::generateCode() {
+  DAWN_LOG(INFO) << "Starting code generation for GTClang ...";
+
+  // Generate code for StencilInstantiations
+  std::map<std::string, std::string> stencils;
+  for(const auto& nameStencilCtxPair : context_->getStencilInstantiationMap()) {
+    std::shared_ptr<iir::StencilInstantiation> origSI = nameStencilCtxPair.second;
+    // TODO the clone seems to be broken
+    //    std::shared_ptr<iir::StencilInstantiation> stencilInstantiation = origSI->clone();
```

Comment:
rm

Reply:
ah yes, this one I would like to keep, because here precisely we would like to clone, rather than inline the current IIR (since we might want to use the original IIR later)
However it depends on your work as well, so we can do it once we merge your stuff